src/main/java/ladder/domain/LadderPartFactory.java

```diff
@@ -1,30 +1,31 @@
 package ladder.domain;
 
+import ladder.strategy.MakeRungStrategy;
+import ladder.util.function.Function;
+
 import java.util.Random;
-import java.util.function.Predicate;
 
 public class LadderPartFactory {
 
-    private static final Predicate<Integer> RAIL_CHECK = idx -> idx % 2 == 0;
+    private static final Random RANDOM = new Random();
+    private static boolean isPreviousRungSet = false;
 
-    public static LadderPart ladderPart(int idx, PreviousRung previousRung) {
-        return ladderPart(idx, previousRung.isSet());
-    }
+    private LadderPartFactory() {}
 
-    public static LadderPart ladderPart(int idx, boolean isRungSet) {
-        if (RAIL_CHECK.test(idx)) {
-            return Rail.rail();
-        }
-
-        if (!isRungSet) {
-            return Rung.rung(randomBoolean());
+    public static LadderPart ladderPart(int idx) {
+        if (Function.EVEN_NUMBER.test(idx)) {
+            return Rail.instance();
         }
+        changeIsPreviousRungSet(MakeRungStrategy.of(isPreviousRungSet, RANDOM.nextBoolean()).test());
+        return Rung.from(isPreviousRungSet);
+    }
```

Comment:
전략패턴을 쓰려고 하는 시도는 좋지만, 코드의 가독성은 떨어지고 생성 로직중 이전LadderPart의 논리값 분기는 하지만 랜덤으로 값을 만든다는건 팩토리클래스에서 그냥 정의하고있습니다. 
완전히 유연한 전략패턴이 아니죠. 

```suggestion
    public static LadderPart ladderPart(int idx, MakeRungStrategy strategy) {
        if (Function.EVEN_NUMBER.test(idx)) {
            return Rail.instance();
        }

        if (isPreviousRungSet) {
            changeIsPreviousRungSet(false);
            return Rung.from(isPreviousRungSet);
        }
        
        return Rung.from(strategy.test());
    }
```
저라면 이런식으로 생각해볼수도 있다고 생각해요 ㅎ 

그리고 MakeRungStrategy 인터페이스의 구현체로 랜덤 발판 생성기를 하나 만드는거죠 ㅎ

```java
public class RandomRungSupplier implements MakeRungStrategy{
    private static final Random RANDOM = new Random();
    
    @Override
    public boolean test() {
        return RANDOM.nextBoolean();
    }
}
```

Reply:
아 Rung에게 MakeRungStrategy를 주입하라고 말씀하신 줄 알았었어요. 그래서 Rung이 해당 strategy를 알 필요가 있나? 하는 생각에 LadderPartFactory에다가 놔야겠다고 생각하고 구현하다보니 이렇게 나왔는데, 그게 아니었군요!

제가 import에 대해서 뭔가 잘못된 생각을 가지고 있는 것 같은데요. Line에서 LadderPartFactory.ladderPart(index, new RandomRungSupplier())를 호출하니깐 import xxx.xxx.RandomRungSupplier;가 생기게 되는데, 이렇게 import를 하게 되는 것에 대해 거부감이 생기더라구요. RandomRungSupplier가 없게 되면 Line에서 에러가 나게 되니깐요.
제가 어떤 식으로 생각을 바꿔야할까요?? 아니면 추천해주실만한 서적이나 자료 같은게 있을까요?

Reply:
@soosue 

 이건 이펙티브자바의 [아이템 5](https://catsbi.oopy.io/d7f3a636-b613-453b-91c7-655d71fda2b1#99fc646f-2f60-43d4-9418-07544af02d43)를 참고해보시면 좋을 것 같아요.

구현체를 어디서 주입해줄지에 대해서에 대한 고민이신것 같은데, 그 선은 윤수님이 잘 정해주셔야 할 것 같네요.
개인적으로는 서비스레이어에서 하거나 아예 DI만 전담할 설정영역을 만드는것도 방법이겠죠. 
결국 `SOLID`에 대해 고민하면서 나올만한 주제인데, 현재 해당 사다리 프로젝트는 적절한 서비스계층이 존재하지 않다보니 이 RandomRungSupplier라는 인터페이스가 아닌 구현체를 어디서 의존할지에 대한 고민을 하게되는 것 같아요 ㅎㅎ

개인적으로는 LadderPartFactory에서 해당 구현체에 대한 의존성을 가져도 될 것 같긴하네요. 
그렇다면 이렇게 코드를 작성할 수 있겠죠
```java
public class LadderPartFactory {
    
    private boolean isPreviousRungSet;
    private final MakeRungStrategy makeRungStrategy;

    public LadderPartFactory(MakeRungStrategy makeRungStrategy) {
        this.isPreviousRungSet = false;
        this.makeRungStrategy = makeRungStrategy;
    }

    public LadderPart ladderPart(int idx) {
        if (Function.EVEN_NUMBER.test(idx)) {
            return Rail.instance();
        }
        if (isPreviousRungSet) {
            changeIsPreviousRungSet(false);
            return Rung.from(isPreviousRungSet);
        }

        return Rung.from(makeRungStrategy.test());
    }
    //logic...
}
```